### PR TITLE
adding polling functionality based on threads

### PR DESCRIPTION
**Pull request by disi33**
Adding polling feature:
Users can create a poll with a "@theorem poll [message]" command. Theorem will then start a thread and @ the channel. Each reply to the thread is being scanned for emojis. 
Every emoji found is gonna be added as a reaction to the starting message in the thread as a vote button.

Also handled an error case where the current framework wasn't able to handle reaction notifications.

### CODE REVIEW

File: Middleware/MiddlewarePipeline.cs [hunk 1]
```diff
@@ -21,6 +21,13 @@ public MiddlewarePipeline(SlackProvider slackProvider, IEnumerable<Middleware> m
 
         private void NewMessage(object sender, MessageEventModel message)
         {
+            if(message.Text == null)
```
Comment (haydenmc) [inline Middleware/MiddlewarePipeline.cs]: #5

File: Providers/SlackProvider.cs [hunk 2]
```diff
@@ -290,6 +290,27 @@ public ChannelModel GetChannelByName(string channelName)
             }
         }
 
+        /// <summary>
```
Comment (haydenmc) [inline Providers/SlackProvider.cs]: Update comment

File: Models/Events/MessageEventModel.cs [hunk 3]
```diff
@@ -8,14 +8,34 @@ namespace Theorem.Models.Events
...
     {
+        private float threadId, timeSent;
+
         [JsonProperty("text")]
         public string Text { get; set; }
         
         [JsonProperty("ts")]
-        [JsonConverter(typeof(UnixDateTimeConverter))]
-        public DateTime TimeSent { get; set; }
+        public string TimeSentNative { get; set; }
```
Comment (haydenmc) [inline Models/Events/MessageEventModel.cs]: If we're changing our logic here to store Slack's native `TimeSent`, we should update the migration to drop the old `TimeSent` column, and rename `TimeSentNative` to `SlackTimeSent` to match the convention in `EventModel`.

`ThreadIDNative` -> `SlackThreadId`

Does anyone actually care about the `ThreadID` DateTime value? (`ThreadIDNative`). Is this any different than `TimeSent` with respect to Date+Time ? If not, we should remove `ThreadID`.

`TeamID` -> `TeamId` or `SlackTeamId` if this is a Slack identifier.

File: Providers/SlackProvider.cs [hunk 4]
```diff
@@ -290,6 +290,27 @@ public ChannelModel GetChannelByName(string channelName)
...
+            {
+                httpClient.BaseAddress = new Uri(BaseApiUrl);
+                var postData = new FormUrlEncodedContent(new[] { 
+                    new KeyValuePair<string, string>("token", _apiToken), 
+                    new KeyValuePair<string, string>("channel", message.Channel.SlackId),
+                    new KeyValuePair<string, string>("timestamp", message.TimeSentNative), 
+                    new KeyValuePair<string, string>("name", reaction)
+                }); 
+                var result = await httpClient.PostAsync("reactions.add", postData);
+                // TODO: Parse result, handle errors, retry, etc.
```
Comment (haydenmc) [inline Providers/SlackProvider.cs]: Add GitHub issue

Comment (haydenmc) [reply Providers/SlackProvider.cs]: (for the TODO item)

File: Models/Events/MessageEventModel.cs (hunk 3 above)
Comment (haydenmc) [reply Models/Events/MessageEventModel.cs]: Also wonder if it's worth storing a converted `DateTime` object on insertion to make querying easier (DateTime is easier to deal with in Linq than a string). We should probably have a way to do this, especially if we're not storing the `TimeSent` value as a number for comparison queries (lt, gt, etc).

File: Models/Events/MessageEventModel.cs [hunk 5]
```diff
@@ -12,10 +12,12 @@ public class MessageEventModel : EventModel
...
-        [JsonConverter(typeof(UnixDateTimeConverter))]
-        public DateTime TimeSent { get; set; }
+        public decimal SlackTimeSent { get; set; }
+        
+        [JsonProperty("thread_ts")]
+        public decimal SlackThreadID { get; set; }
         
         [JsonProperty("team")]
-        public string TeamId { get; set; }
+        public string TeamID { get; set; }
```
Comment (haydenmc) [inline Models/Events/MessageEventModel.cs]: `TeamID` -> `SlackTeamId`

File: Models/Events/MessageEventModel.cs [hunk 6]
```diff
@@ -12,10 +12,12 @@ public class MessageEventModel : EventModel
         public string Text { get; set; }
         
         [JsonProperty("ts")]
-        [JsonConverter(typeof(UnixDateTimeConverter))]
-        public DateTime TimeSent { get; set; }
+        public decimal SlackTimeSent { get; set; }
+        
+        [JsonProperty("thread_ts")]
+        public decimal SlackThreadID { get; set; }
```
Comment (haydenmc) [inline Models/Events/MessageEventModel.cs]: `SlackThreadID` -> `SlackThreadId`

Do we need the additional precision that comes with `decimal`? Is `double` good enough?

File: Migrations/ApplicationDbContextModelSnapshot.cs [hunk 7]
```diff
@@ -193,6 +193,12 @@ protected override void BuildModel(ModelBuilder modelBuilder)
 
                     b.Property<DateTime>("TimeSent");
 
+                    b.Property<string>("TimeSentNative");
```
Comment (haydenmc) [inline Migrations/ApplicationDbContextModelSnapshot.cs]: This should probably be regenerated - these properties (`TimeSentNative`, `ThreadID`, `ThreadIDNative`) are out of date.

File: Models/Events/MessageEventModel.cs (hunk 6 above)
Comment (disi33) [reply Models/Events/MessageEventModel.cs]: As discussed offline, we need the precision provided by decimal

**Pull request merged**